File: Models/Bloque.php
<?php

class Bloque
{
    private $PDO;

    public function __construct()
    {
        require_once __DIR__ . '/../config/conexion.php';
        $con = new Database();
        $this->PDO = $con->conexion();
        date_default_timezone_set('America/Bogota');
    }
    

    public function obtenerUltimoHash()
    {
        $sql = "SELECT hash FROM blockchain ORDER BY id DESC LIMIT 1";
        $stmt = $this->PDO->query($sql);
        $fila = $stmt->fetch(PDO::FETCH_ASSOC);
        return $fila['hash'] ?? '0';
    }

    //Función para crear un nuevo usuario
    public function createUser($usuario, $contrasena)
    {
        $sql = "INSERT INTO usuarios (usuario, contrasena) VALUES (:nombre, :contrasena)";
        $stmt = $this->PDO->prepare($sql);
        $stmt->bindParam(':nombre', $usuario);
        $stmt->bindParam(':contrasena', $contrasena);
        if ($stmt->execute()) {
            return $this->PDO->lastInsertId(); // ID del nuevo usuario
        } else {
            return false;
        }
    }

    // Función para insertar crear un nuevo bloque utilizando el id de usuario
    public function createBlo($id_usuario, $estado)
    {
        $anteriorHash = $this->obtenerUltimoHash();
        $fecha = date("Y-m-d H:i:s");
        $datos = $fecha . $estado . $anteriorHash;
        $hash = hash('sha256', $datos);

        $sql = "INSERT INTO blockchain (id_usuario, fecha, estado, anterior_hash, hash)
                VALUES (:id_usuario, :fecha, :estado, :anterior_hash, :hash)";
        $stmt = $this->PDO->prepare($sql);
        $stmt->bindParam(':id_usuario', $id_usuario);
        $stmt->bindParam(':fecha', $fecha);
        $stmt->bindParam(':estado', $estado);
        $stmt->bindParam(':anterior_hash', $anteriorHash);
        $stmt->bindParam(':hash', $hash);

        return $stmt->execute();
    }

    // Función para iniciar sesión
    public function login($usuario, $contrasena)
    {
        $stmt = $this->PDO->prepare("SELECT * FROM usuarios WHERE usuario = :usuario AND contrasena = :contrasena");
        $stmt->bindParam(':usuario', $usuario);
        $stmt->bindParam(':contrasena', $contrasena);
        $stmt->execute();
        return ($stmt->rowCount() > 0) ? $stmt->fetch() : false;
  
    }
}

    

File: Models/Bloque.py
from datetime import datetime
import hashlib
from models.database import Database

class Bloque:
    def __init__(self):
        self.db = Database()
    
    
    #Obtiene el último hash para agregarlo al historial
    def obtener_ultimo_hash(self):
        conn = self.db.conexion()
        cursor = conn.cursor(dictionary=True)
        
        sql = "SELECT hash FROM blockchain ORDER BY id DESC LIMIT 1"
        cursor.execute(sql)
        fila = cursor.fetchone()
        
        cursor.close()
        conn.close()
        
        return fila['hash'] if fila else '0'
    
    #Función para crear un nuevo usuario
    def create_user(self, usuario, contrasena):
        if self.usuario_existe(usuario):
            return "Usuario ya registrado"
        
        conn = self.db.conexion()
        cursor = conn.cursor()
        
        sql = "INSERT INTO usuarios (usuario, contrasena) VALUES (%s, %s)"
        cursor.execute(sql, (usuario, contrasena))
        conn.commit()
        
        last_id = cursor.lastrowid
        
        cursor.close()
        conn.close()
        
        return last_id if cursor.rowcount > 0 else False

    #Función para crear un nuevo registro al interactuar con el botón
    def create_blo(self, id_usuario, estado):
        anterior_hash = self.obtener_ultimo_hash()
        fecha = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        datos = fecha + estado + anterior_hash
        hash_value = hashlib.sha256(datos.encode()).hexdigest()
        
        conn = self.db.conexion()
        cursor = conn.cursor()
        
        sql = """INSERT INTO blockchain (id_usuario, fecha, estado, anterior_hash, hash)
                VALUES (%s, %s, %s, %s, %s)"""
        cursor.execute(sql, (id_usuario, fecha, estado, anterior_hash, hash_value))
        conn.commit()
        
        result = cursor.rowcount > 0
        
        cursor.close()
        conn.close()
        
        return result
    
    #Función para iniciar sesión
    def login(self, usuario, contrasena):
        conn = self.db.conexion()
        cursor = conn.cursor(dictionary=True)
        
        sql = "SELECT * FROM usuarios WHERE usuario = %s AND contrasena = %s"
        cursor.execute(sql, (usuario, contrasena))
        user = cursor.fetchone()
        
        cursor.close()
        conn.close()
        
        return user
    
    #Función para verificar la contraseña del usuario 
    def password_veri(self, usuario, contrasena):
        conn = self.db.conexion()
        cursor = conn.cursor(dictionary=True)
        
        sql = "SELECT contrasena FROM usuarios WHERE usuario = %s"
        cursor.execute(sql, (usuario,))
        user = cursor.fetchone()
        
        cursor.close()
        conn.close()
        
        if user and contrasena == user['contrasena']:
            return True
        return False
    
    #Función que verifica que el nombre de usuario no exista en la base de datos
    def usuario_existe(self, usuario):
        conn = self.db.conexion()
        cursor = conn.cursor(dictionary=True, buffered=True)
        
        sql = "SELECT id FROM usuarios WHERE usuario = %s"
        cursor.execute(sql, (usuario,))
        resultado = cursor.fetchone()
        
        cursor.close()
        conn.close()
        
        return resultado is not None
